patches/tModLoader/Terraria/GameContent/Achievements/ItemPickupCondition.cs.patch

```diff
@@ -0,0 +1,48 @@
+--- src/TerrariaNetCore/Terraria/GameContent/Achievements/ItemPickupCondition.cs
++++ src/tModLoader/Terraria/GameContent/Achievements/ItemPickupCondition.cs
+@@ -3,6 +_,9 @@
+ 
+ namespace Terraria.GameContent.Achievements;
+ 
++/// <summary>
++/// Satisfied when any of the provided items is picked up.
++/// </summary>
+ public class ItemPickupCondition : AchievementCondition
+ {
+ 	private const string Identifier = "ITEM_PICKUP";
+@@ -10,8 +_,10 @@
+ 	private static bool _isListenerHooked;
+ 	private short[] _itemIds;
+ 
++	private static string IDToIdentifier(short itemId) => ModLoader.ItemLoader.GetItem(itemId)?.FullName ?? itemId.ToString();
++
+-	private ItemPickupCondition(short itemId)
++	public ItemPickupCondition(short itemId)
```

Comment:
There's no need to make these `public` because there are static factory `Create` methods. Arguably though, the factory methods are noise/bloat/non idiomatic, so if that's why you chose to expose the constructors instead that's fine

Reply:
`ItemPickupCondition.Create` returns a `AchievmentCondition`, not a `ItemPickupCondition`. Same with the other classes. I guess it is technically safe to change their return values (although they are public, but I don't expect anyone to mind that a non-supported method changed return type in an update), but then I'd just have another method in the chain. `ModAchiemvent.AddNPCKilledCondition->AddCondition->NPCKilledCondition.Create->NPCKilledCondition.ctor` instead of `ModAchiemvent.AddNPCKilledCondition->AddCondition->NPCKilledCondition.ctor`.

We should definitely revisit this more, it is quite messy. I have an advanced example I've been working on not yet pushed.